src/v/kafka/server/handlers/produce.cc

```diff
@@ -190,6 +179,31 @@ static partition_produce_stages partition_append(
     };
 }
 
+struct record_batch_pair {
+    model::record_batch batch;
```

Comment:
Does the cover letter address this design. It seemed to have swapped from caching keys to full copy. 

Means at least 2x. Which is ok as long as short lived. One thing to take into account is if we have to put more back pressure on the Kafka layer give That a compacted topic is at least 2x mem ? 

In the semaphore reservation of the RPC it seems like some accounting is missing.